### ISSUE TITLE
Illegal Seek / Mergerfs Question

### ISSUE THREAD

**Issue by animosity22** [rclone/rclone#3461]
#### What is the problem you are having with rclone?

Illegal Seek with mergerfs/rclone

#### What is your rclone version (output from `rclone version`)

1.48

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Ubuntu 18.04

####  Which cloud storage system are you using? (eg Google Drive)

Google Drive

I have the issue fairly well documented and written up here:

https://github.com/trapexit/mergerfs/issues/656

and was hoping to get some clarity on why rclone is working but mergerfs seems to tossing up the Illegal Seek.

**Comment by ncw** [comment]
> I think the different is related to how the file is being replaced on mergerfs compared to just the rclone mount and the seek error and your comment made me test with --vfs-cache-mode writes, which seems to fix the issue as it wants to seek.

I think this is the nub of the problem.

What might be happening is that the writes are re-ordered for some reason so instead of `write 100 bytes to position 0; write 100 bytes to position 100` we are getting `write 100 bytes to position 100; write 100 bytes to position 0`.  This would cause the illegal seek.

I recently put in a work-around for this as it happens in `rclone serve sftp` too. See d377842395b897e9cee96f6a1c2fdd25f20e4f9c and af030f74f5c9c49d337ad768b307e50d3cfd60fe.  The first was in v1.48.0 but the second isn't so it might be worth having a go with [the latest beta](https://beta.rclone.org) to see if it fixes it.

If it is still a problem a log with `-vv` of the problem happening should shed some light.

I don't see any ERRORs in your logs in the other issue - I'd really expect to see an ERROR corresponding to that illegal seek.

**Comment by trapexit** [comment]
If you look at the mergerfs strace you can see that it's a single write starting from offset 0.

**Comment by ncw** [comment]
> If you look at the mergerfs strace you can see that it's a single write starting from offset 0.

I see what you mean.

I just had a look through the rclone source - there isn't anywhere it returns ESPIPE without logging an ERROR first as far as I can see unless you try to seek when `--no-seek` is in use.

None of the calls to rclone from FUSE returned an error you can see that in the logs. The 243 bytes were written correctly and no errors were returned to fuse.

```
2019/08/20 13:09:33 DEBUG : &{hosts (w)}: Write: len=243, offset=0
2019/08/20 13:09:33 DEBUG : &{hosts (w)}: >Write: written=243, err=<nil>
2019/08/20 13:09:33 DEBUG : &{hosts (w)}: Flush:
2019/08/20 13:09:33 DEBUG : Encrypted drive 'gcrypt:': File to upload is small (243 bytes), uploading instead of streaming
2019/08/20 13:09:34 INFO  : hosts: Copied (new)
2019/08/20 13:09:34 DEBUG : &{hosts (w)}: >Flush: err=<nil>
2019/08/20 13:09:34 DEBUG : &{hosts (w)}: Release:
2019/08/20 13:09:34 DEBUG : hosts: WriteFileHandle.Release nothing to do
2019/08/20 13:09:34 DEBUG : &{hosts (w)}: >Release: err=<nil>
```

So where is the ESPIPE coming from?

**Comment by ncw** [comment]
I managed to replicate this problem with

```
mergerfs -o use_ino,allow_other,func.getattr=newest,category.action=all,category.create=ff,dropcacheonclose=true /mnt/local:/mnt/gd /mnt/gmedia
rclone mount -vv drive: /mnt/gd
cp -a ~/go/src/github.com/rclone/rclone/README.md /mnt/gmedia/justaddedfile3.txt
```

> So where is the ESPIPE coming from?

I've discovered that... it is coming directly from FUSE so it never touches rclone.

This is because I've set the can't seek FUSE hint on the file handle.

I think it is failing because FUSE calls Truncate() on the file handle, then calls Write()?

I'm not sure I need to set the hint - rclone does enough checking on its own.  The docs on the [nonseekable flag](https://libfuse.github.io/doxygen/structfuse__file__info.html#a272022c57a6a79dd8f98ef597786e154) aren't exactly comprehensive!

@trapexit - do you have an opinion?

@animosity22 - can you try this patch and see if it does the same for you? I don't think I'm using the current mergerfs I'm using that packaged with ubuntu 19.04

binary attached: [rclone.gz](https://github.com/rclone/rclone/files/3526061/rclone.gz)


```patch
diff --git a/cmd/mount/file.go b/cmd/mount/file.go
index 843d5fff1..0236aca97 100644
--- a/cmd/mount/file.go
+++ b/cmd/mount/file.go
@@ -4,7 +4,6 @@ package mount
 
 import (
 	"context"
-	"io"
 	"time"
 
 	"bazil.org/fuse"
@@ -75,9 +74,9 @@ func (f *File) Open(ctx context.Context, req *fuse.OpenRequest, resp *fuse.OpenR
 	}
 
 	// See if seeking is supported and set FUSE hint accordingly
-	if _, err = handle.Seek(0, io.SeekCurrent); err != nil {
-		resp.Flags |= fuse.OpenNonSeekable
-	}
+	// if _, err = handle.Seek(0, io.SeekCurrent); err != nil {
+	// 	resp.Flags |= fuse.OpenNonSeekable
+	// }
 
 	return &FileHandle{handle}, nil
 }
```

**Comment by animosity22** [comment]
@ncw  - That works for me.

I just ran the same copy test and it worked without issue.

Here is the debug output from the rclone mount and the copy worked.


```

2019/08/21 11:30:23 DEBUG : /: Attr:
2019/08/21 11:30:23 DEBUG : /: >Attr: attr=valid=1s ino=0 size=0 mode=drwxrwxr-x, err=<nil>
2019/08/21 11:30:23 DEBUG : hosts: Attr:
2019/08/21 11:30:23 DEBUG : hosts: >Attr: a=valid=1s ino=0 size=276 mode=-rw-rw-r--, err=<nil>
2019/08/21 11:30:23 DEBUG : hosts: Open: flags=OpenWriteOnly
2019/08/21 11:30:23 DEBUG : hosts: Open: flags=O_WRONLY
2019/08/21 11:30:23 DEBUG : hosts: >Open: fd=hosts (w), err=<nil>
2019/08/21 11:30:23 DEBUG : hosts: >Open: fh=&{hosts (w)}, err=<nil>
2019/08/21 11:30:23 DEBUG : hosts: Setattr: a=Setattr [ID=0x46 Node=0x5 Uid=1000 Gid=1000 Pid=3572] size=0 handle=INVALID-0x0 lockowner
2019/08/21 11:30:23 DEBUG : hosts: Truncating 1 file handles
2019/08/21 11:30:23 DEBUG : hosts: >Setattr: err=<nil>
2019/08/21 11:30:23 DEBUG : hosts: Attr:
2019/08/21 11:30:23 DEBUG : hosts: >Attr: a=valid=1s ino=0 size=276 mode=-rw-rw-r--, err=<nil>
2019/08/21 11:30:23 DEBUG : &{hosts (w)}: Write: len=276, offset=0
2019/08/21 11:30:23 DEBUG : &{hosts (w)}: >Write: written=276, err=<nil>
2019/08/21 11:30:23 DEBUG : &{hosts (w)}: Flush:
2019/08/21 11:30:23 DEBUG : Encrypted drive 'gcrypt:': File to upload is small (276 bytes), uploading instead of streaming
2019/08/21 11:30:24 INFO  : hosts: Copied (new)
2019/08/21 11:30:24 DEBUG : &{hosts (w)}: >Flush: err=<nil>
2019/08/21 11:30:24 DEBUG : &{hosts (w)}: Flush:
2019/08/21 11:30:24 DEBUG : hosts: WriteFileHandle.Flush nothing to do
2019/08/21 11:30:24 DEBUG : &{hosts (w)}: >Flush: err=<nil>
2019/08/21 11:30:24 DEBUG : &{hosts (w)}: Release:
2019/08/21 11:30:24 DEBUG : hosts: WriteFileHandle.Release nothing to do
2019/08/21 11:30:24 DEBUG : &{hosts (w)}: >Release: err=<nil>
```

**Comment by trapexit** [comment]
https://lkml.org/lkml/2019/6/9/183

nonseekable is a bit broken AFAIU anyway.

Probably better for rclone to manage the situation since it's really not a streaming file so much as a limitation. Not sure if there are secondary affects of labeling it as such so not doing so gives rclone more control over the situation.

**Comment by ncw** [comment]
I had a quick troll through the linux kernel source.  It looks like if **any** seeking is done on a nonseekable file (even if it seeks to the same place) it will fail with ESPIPE.

**Comment by ncw** [comment]
Here is a beta with that fix - can you give it a go @animosity22 ?

https://beta.rclone.org/branch/v1.48.0-212-g81cde2b4-fix-3461-seek-beta/ (uploaded in 15-30 mins)

Thanks for the analysis @trapexit - I can't think of a reason for setting the hint any more so I've unset it.

**Comment by ncw** [comment]
Thanks for comments and testing

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.49

**Comment by animosity22** [comment]
Looks good! Thanks!
